### DISCUSSION TITLE
iOS and Android native code reuse?

### DISCUSSION THREAD

**Discussion by grahammendick** [reactwg/react-native-new-architecture#24] [Q&A] [answered]
Thanks to [your step by step guide](https://github.com/cipolleschi/RNNewArchitectureLibraries/blob/feat/back-fabric-comp/RUN.md), I’m slowly getting my head around the changes needed to port a native component to the new architecture on iOS. It looks like I can’t reuse any of my existing Objective-C View code. I have to copy it over into the new ComponentView.mm class. Is that right? Is it the same story on Android or can I expect to reuse my existing Managers and Views?

**Comment by cipolleschi** [comment, marked as answer]
Hi @grahammendick!

We are about to migrate to the community the repo you linked, which we updated with the Android side of things just today.
As you will see, for Android the proposed approach is to have two different sourceset which instanciate and use a common implementation.  

So, to answer your question, you still have to create different files, but with a bit of engineering, you can keep using your Android code for both the Old and the New Architecture.

Let me know if I can help you with anything else!

**Reply by grahammendick** [reply]
Thanks. It's great to have a step by step guide for Android, too.

Is the React Native source code using 2 different sourcesets? It looks like there's only one manager, for example, `ReactModalHostManager`.

**Reply by cortinico** [reply]
Hi @grahammendick, Thanks for raising this question. @cipolleschi already answered correctly: the suggested approach is to use two source sets. The reason behind this is that it makes for more flexibility and allows you, as a library author, to support more users also on previous version of RN.

The rationale behind using two source sets is that the New Architecture relies on the Codegen. The codegen is executed inside the React Native Gradle Plugin which is provided by the host app (as a `classpath` dependency in the top-level `build.gradle`).

You might want your library to work also for users which don't apply the plugin yet. Therefore you need 2 source sets as those users won't generate the class your code is extending from. 

> Is the React Native source code using 2 different sourcesets? It looks like there's only one manager, for example, `ReactModalHostManager`.

Nope, the RN core components are not using 2 source sets. The reason behind this is that RN is shipped as a prebuilt on Android (i.e. is an `.aar`) inside the `/android` folder of the NPM package. This allows us to ignore the problem of having (or not) the codegen for our users as all the code is codegen-ed and build at publish time. This is another approach you could take, but is not the one we're suggesting, as you'll also need to ship a native library (a `.so`) and we're suggesting a full build from source to avoid ABI incompatibilities problems.

Hope this clarifies your doubts.

**Reply by grahammendick** [reply]
Aha, thanks, this makes senes
